### PR TITLE
golang: update to 1.22.1

### PR DESCRIPTION
**Pull request by mochaaP**
Maintainer: @jefferyto 
Compile tested: `aarch64_cortex-a53` @ SNAPSHOT
Run tested: `x86_64-generic` @ SNAPSHOT

Description:
Updated to go 1.22.1

Added a third bootstrap stage (go1.21.5, the last known-good build), since go1.22 (and onwards) requires at least go1.20.14 to build.[^1]

Fixes: https://github.com/openwrt/packages/issues/23541

[^1]: https://go.dev/doc/go1.22#bootstrap

### CODE REVIEW

**Comment by predators46** [comment]
I like this. tested openwrt 21.02.7

**Comment by predators46** [comment]
@mochaaP 

the problem occurs if building the sing-box package using openwrt-sdk. but if you build sing-box on the device after installing the golang package on openwrt sing-box can be made ```-linkmode requires external (cgo) linking, but cgo is not enabled```

**Comment by mochaaP** [comment]
> @mochaaP 
> 
> the problem occurs if building the sing-box package using openwrt-sdk. but if you build sing-box on the device after installing the golang package on openwrt sing-box can be made 
> 
> ```-linkmode requires external (cgo) linking, but cgo is not enabled```

You need `CGO_ENABLED=1`

**Comment by predators46** [comment]
> > @mochaaP
> > the problem occurs if building the sing-box package using openwrt-sdk. but if you build sing-box on the device after installing the golang package on openwrt sing-box can be made
> > `-linkmode requires external (cgo) linking, but cgo is not enabled`
> 
> You need `CGO_ENABLED=1`

https://github.com/golang/go/issues/65887

**Comment by predators46** [comment]
the problem is in the Makefile I am using. now everything is running smoothly

**Comment by mochaaP** [comment]
@1715173329

**Comment by BKPepe** [comment]
> I like this. tested openwrt 21.02.7

Are you sure with this OpenWrt version? This needs to be compile and run tested for the master branch. Not on deprecated and already end of life branch.

**Comment by predators46** [comment]
> > I like this. tested openwrt 21.02.7
> 
> Are you sure with this OpenWrt version? This needs to be compile and run tested for the master branch. Not on deprecated and already end of life branch.



> > I like this. tested openwrt 21.02.7
> 
> Are you sure with this OpenWrt version? This needs to be compile and run tested for the master branch. Not on deprecated and already end of life branch.

I have tried openwrt 23.05.2. the master branch should have no problems

File: lang/golang/golang/Makefile [hunk 1]
```diff
@@ -95,6 +95,12 @@ BOOTSTRAP_1_17_HASH:=a1a48b23afb206f95e7bbaa9b898d965f90826f6f1d1fc0c1d784ada0cd
 
 BOOTSTRAP_1_17_BUILD_DIR:=$(HOST_BUILD_DIR)/.go_bootstrap_1.17
 
+BOOTSTRAP_1_21_SOURCE:=go1.21.5.src.tar.gz
```
Comment (1715173329) [inline lang/golang/golang/Makefile:98]: basically we would use the oldest go toolchain for bootstrap, i.e. go1.20.15.

Comment (mochaaP) [reply lang/golang/golang/Makefile:98]: Since this is already verified on OpenWrt and does not require any patches on our side (iirc 1.20 needs some patches), this should be good.

**Comment by predators46** [comment]
version go1.22.1.src.tar.gz
bootstrap version go1.20.14.src.tar.gz 

all in all no problem when building golang packages

File: lang/golang/golang/Makefile [hunk 2]
```diff
@@ -104,6 +110,7 @@ PKG_UNPACK:=$(HOST_TAR) -C "$(PKG_BUILD_DIR)" --strip-components=1 -xzf "$(DL_DI
 HOST_UNPACK:=$(HOST_TAR) -C "$(HOST_BUILD_DIR)" --strip-components=1 -xzf "$(DL_DIR)/$(PKG_SOURCE)"
 BOOTSTRAP_UNPACK:=$(HOST_TAR) -C "$(BOOTSTRAP_BUILD_DIR)" --strip-components=1 -xzf "$(DL_DIR)/$(BOOTSTRAP_SOURCE)"
 BOOTSTRAP_1_17_UNPACK:=$(HOST_TAR) -C "$(BOOTSTRAP_1_17_BUILD_DIR)" --strip-components=1 -xzf "$(DL_DIR)/$(BOOTSTRAP_1_17_SOURCE)"
```
Comment (BKPepe) [inline lang/golang/golang/Makefile:112]: Is there any reason, why do we need to keep bootstrap 1.17 and why we could not use the 1.21, which you added?

**Comment by BKPepe** [comment]
Commit with the update 1.22.1 isnt sufficient, because you are missing commit description. Please take a look at the Git history, how @jefferyto and others were doing it.

File: lang/golang/golang/Makefile (hunk 2 above)
Comment (mochaaP) [reply lang/golang/golang/Makefile:112]: This was already stated in the 1.22.0 commit msg.

tl;dr: golang 1.22+ needs at least go1.20.15 to compile

File: lang/golang/golang/Makefile [hunk 3]
```diff
@@ -251,8 +274,13 @@ define Host/Compile
 		$(HOST_GO_VARS) \
 	)
 
-	$(call GoCompiler/Host/Make, \
+	$(call GoCompiler/Bootstrap-1.21/Make, \
 		GOROOT_BOOTSTRAP="$(BOOTSTRAP_1_17_BUILD_DIR)" \
```
[the remark this answers is not in the document]
Comment (mochaaP) [reply lang/golang/golang/Makefile:278]: see the comments below (https://github.com/openwrt/packages/pull/23475#discussion_r1518812323)

**Comment by mochaaP** [comment]
@BKPepe:
> Commit with the update 1.22.1 isnt sufficient, because you are missing commit description.

Done. [`7ce9df6` (#23475)](https://github.com/openwrt/packages/pull/23475/commits/7ce9df619e46aca23689c30872cbbf11d0eda84c)

File: lang/golang/golang/Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (BKPepe) [reply lang/golang/golang/Makefile:278]: I think it does not help me. You call Bootstrap 1.21 and then the goroot buildir is 1.17. That does not seems to be intentional or is it? Still all Go lang packages requirest up to date Go, so having three bootstrap does not seems right. Why do we need to keep go1.4-bootstrap-20171003.tar.gz?

[the remark this answers is not in the document]
Comment (mochaaP) [reply lang/golang/golang/Makefile:278]: This is intentional. The `GOROOT_BOOTSTRAP` is for specifying the previous built go toolchain, to bootstrap the next stage.

go1.4 is required for bootstrapping because it's the last version of go with its compiler written in C.

[the remark this answers is not in the document]
Comment (BKPepe) [reply lang/golang/golang/Makefile:278]: Oh, you are right. I learned and checked how the Buildroot does it and there are those two stages. Seems complicated, anyway thanks for your time!

File: lang/golang/golang/Makefile (hunk 1 above)
Comment (1715173329) [reply lang/golang/golang/Makefile:98]: > iirc 1.20 needs some patches

Not really, we used that version in the past.

**Comment by xh0823** [comment]
When can it be merged? Is there any error?

**Comment by 1715173329** [comment]
As an old man I still vote to use go 1.20 for bootstrap, otherwise LGTM. @jefferyto might be on vacation?

**Review by 1715173329 — APPROVED** [review]
